Batch close-indices cluster state updates

joegallo opened the pull request:
Related to https://github.com/elastic/elasticsearch/issues/81627 and https://github.com/elastic/elasticsearch/pull/83432, followup to https://github.com/elastic/elasticsearch/pull/83760.

joegallo committed:
Switch to a (non-static) inner class

joegallo committed:
Drop this variable

joegallo committed:
Drop this variable

joegallo committed:
Extract an executor for adding blocks

joegallo committed:
Cleanup some crusty collections code

joegallo committed:
Fix typos

joegallo committed:
Build the listener inside the executor

rather than passing it in

joegallo committed:
Do the same for the close-indices task

elasticsearchmachine commented:
Hi @joegallo, I've created a changelog YAML for you.

joegallo committed:
Update docs/changelog/84259.yaml

DaveCTurner commented on server/src/main/java/org/elasticsearch/cluster/metadata/MetadataIndexStateService.java:
```diff
@@ -1023,24 +1088,7 @@ public static ClusterBlock createUUIDBasedBlock(ClusterBlock clusterBlock) {
-            this.indicesService = indicesService;
-            this.shardLimitValidator = shardLimitValidator;
-        }
+    private class OpenIndicesExecutor implements ClusterStateTaskExecutor<OpenIndicesTask> {
```
Slight preference to keep this (and other executors) as static classes that hold onto the services they need directly.

joegallo replied:
Ehhhh, this is a bit of a pain -- `private class WaitForClosedBlocksApplied` / ` private class WaitForBlocksApplied`, and the former is invoked by `AddBlocksToCloseExecutor`, so I'd have to walk those two back, too.

I'm really with you on disliking (non-static) inner classes, they feel a bit too ['more magic'](http://www.catb.org/jargon/html/magic-story.html) for me, but in this case they're all private anyway, and the juggling of the fields across them all just seems like noise.

That said, this isn't 'no', it's just 'ehhh, this is ever so slightly harder than you might have thought' -- feel free to say something like "yeah, that's cool, please make the change anyway" and I will happily do it.

DaveCTurner replied:
My preference is only slight, I'll defer to your judgement.

DaveCTurner commented on server/src/main/java/org/elasticsearch/cluster/metadata/MetadataIndexStateService.java:
```diff
@@ -147,114 +157,170 @@ public MetadataIndexStateService(
+                    indices.addAll(closingResult.v2());
+                    state = closingResult.v1();
+
+                    builder.success(task, new ActionListener<>() {
```
Suggest `task.listener().delegateFailure(...)` here.

joegallo replied:
053c762cfcdbb77c79a1589a6e957cd17dcef423

DaveCTurner commented on server/src/main/java/org/elasticsearch/cluster/metadata/MetadataIndexStateService.java:
```diff
@@ -147,114 +157,170 @@ public MetadataIndexStateService(
+
+                    Tuple<ClusterState, Collection<IndexResult>> closingResult = closeRoutingTable(state, blockedIndices, verifyResults);
+                    assert verifyResults.size() == closingResult.v2().size();
+                    indices.addAll(closingResult.v2());
```
Do we need to copy `closingResult.v2()` to a new list here? Think we could just use it directly in the publish listener.

joegallo replied:
d3445d2a27bad2477845c81d794c205a80f00595 and then d64a1037c7dac962bc309ea49fed212ffa0e7854

DaveCTurner commented on server/src/main/java/org/elasticsearch/cluster/metadata/MetadataIndexStateService.java:
```diff
@@ -147,114 +157,170 @@ public MetadataIndexStateService(
+                                        new WaitForClosedBlocksApplied(
+                                            blockedIndices,
+                                            task.request,
+                                            ActionListener.wrap(
```
and here, `task.listener().delegateFailure(...)`.

DaveCTurner commented on server/src/main/java/org/elasticsearch/cluster/metadata/MetadataIndexStateService.java:
```diff
@@ -147,114 +157,170 @@ public MetadataIndexStateService(
+                try {
+                    final Map<Index, ClusterBlock> blockedIndices = new HashMap<>(task.request.indices().length);
+                    state = addIndexClosedBlocks(task.request.indices(), blockedIndices, state);
+                    builder.success(task, new ActionListener<>() {
```
and here, `task.listener().delegateFailure(...)`.

DaveCTurner reviewed (commented):
Looks good, I left a few nits but seems structurally sound to me.

DaveCTurner commented on server/src/main/java/org/elasticsearch/cluster/metadata/MetadataIndexStateService.java:
```diff
@@ -147,114 +157,170 @@ public MetadataIndexStateService(
+                            if (blockedIndices.isEmpty()) {
+                                task.listener.onResponse(new CloseIndexResponse(true, false, Collections.emptyList()));
+                            } else {
+                                threadPool.executor(ThreadPool.Names.MANAGEMENT)
```
I don't see a great reason to fork here, we really just trigger a bunch of transport actions that themselves fork to `MANAGEMENT` again. But then again this is sending `O(#shards)` transport actions in parallel, which could be lots, and we know that can cause problems. No change needed here in this PR but we might want to do something about that in future perhaps.

joegallo committed:
Merge branch 'master' into batch-add-blocks-and-close

joegallo committed:
Switch to task.listener.delegateFailure(...)

joegallo committed:
Cleanup some more crusty collections code

joegallo committed:
Clean this up a bit to avoid the addAll

DaveCTurner commented on server/src/main/java/org/elasticsearch/cluster/metadata/MetadataIndexStateService.java:
```diff
@@ -147,114 +157,153 @@ public MetadataIndexStateService(
+                    state = addIndexClosedBlocks(task.request.indices(), blockedIndices, state);
+                    builder.success(task, task.listener.delegateFailure((l1, clusterState) -> {
+                        if (blockedIndices.isEmpty()) {
+                            task.listener.onResponse(new CloseIndexResponse(true, false, Collections.emptyList()));
```
Should really use the delegated listener here, although secretly it's the listener we first thought of:

```suggestion
                            l1.onResponse(new CloseIndexResponse(true, false, Collections.emptyList()));
```

DaveCTurner commented on server/src/main/java/org/elasticsearch/cluster/metadata/MetadataIndexStateService.java:
```diff
@@ -147,114 +157,153 @@ public MetadataIndexStateService(
+                                    new WaitForClosedBlocksApplied(
+                                        blockedIndices,
+                                        task.request,
+                                        task.listener.delegateFailure((l2, verifyResults) -> {
```
```suggestion
                                        l1.delegateFailure((l2, verifyResults) -> {
```

joegallo replied:
1485315cc6ef89c399c142daec05140b8a075794

DaveCTurner commented on server/src/main/java/org/elasticsearch/cluster/metadata/MetadataIndexStateService.java:
```diff
@@ -147,114 +157,153 @@ public MetadataIndexStateService(
+                                        task.listener.delegateFailure((l2, verifyResults) -> {
+                                            clusterService.submitStateUpdateTask(
+                                                "close-indices",
+                                                new CloseIndicesTask(task.request, blockedIndices, verifyResults, task.listener),
```
```suggestion
                                                new CloseIndicesTask(task.request, blockedIndices, verifyResults, l2),
```

joegallo replied:
1485315cc6ef89c399c142daec05140b8a075794

DaveCTurner commented on server/src/main/java/org/elasticsearch/cluster/metadata/MetadataIndexStateService.java:
```diff
@@ -147,114 +157,153 @@ public MetadataIndexStateService(
+                                    // so we maintain a kind of coherency by overriding the shardsAcknowledged value
+                                    // (see ShardsAcknowledgedResponse constructor)
+                                    boolean shardsAcked = acknowledged ? shardsAcknowledged : false;
+                                    task.listener.onResponse(new CloseIndexResponse(acknowledged, shardsAcked, indices));
```
```suggestion
                                    l.onResponse(new CloseIndexResponse(acknowledged, shardsAcked, indices));
```

joegallo replied:
1485315cc6ef89c399c142daec05140b8a075794

DaveCTurner commented on server/src/main/java/org/elasticsearch/cluster/metadata/MetadataIndexStateService.java:
```diff
@@ -147,114 +157,153 @@ public MetadataIndexStateService(
+                                    boolean shardsAcked = acknowledged ? shardsAcknowledged : false;
+                                    task.listener.onResponse(new CloseIndexResponse(acknowledged, shardsAcked, indices));
+                                },
+                                task.listener::onFailure
```
```suggestion
                                l::onFailure
```

joegallo replied:
1485315cc6ef89c399c142daec05140b8a075794

DaveCTurner commented on server/src/main/java/org/elasticsearch/cluster/metadata/MetadataIndexStateService.java:
```diff
@@ -147,114 +157,153 @@ public MetadataIndexStateService(
                             );
-                    }
+                        } else {
+                            task.listener.onResponse(new CloseIndexResponse(acknowledged, false, indices));
```
```suggestion
                            l.onResponse(new CloseIndexResponse(acknowledged, false, indices));
```

joegallo replied:
1485315cc6ef89c399c142daec05140b8a075794

DaveCTurner reviewed (commented):
Nits

joegallo committed:
Change the return type

Since we're just building a list with the returned collection anyway,
we might as well internalize that since this is the only production
invocation anyway.

Changed finalizeBlock, too, for parity, although that doesn't have the
full benefit here just yet because it hasn't been task-ified.

joegallo committed:
Use the passed-in listener

DaveCTurner reviewed (approved):
LGTM

original-brownbear commented on server/src/main/java/org/elasticsearch/cluster/metadata/MetadataIndexStateService.java:
```diff
@@ -460,7 +505,7 @@ public ClusterState execute(final ClusterState currentState) {
                     if (oldState == newState) {
                         assert blockedIndices.isEmpty() : "List of blocked indices is not empty but cluster state wasn't changed";
-                        listener.onResponse(new AddIndexBlockResponse(true, false, Collections.emptyList()));
+                        listener.onResponse(new AddIndexBlockResponse(true, false, List.of()));
```
NIT: same here, if we're changing this line anyway we could make the empty response a constant

joegallo replied:
aa32395

original-brownbear reviewed (approved):
LGTM (nits really can be ignored, just had to let them out :))

joegallo committed:
Extract empty responses as constants

joegallo committed:
Merge branch 'master' into batch-add-blocks-and-close

joegallo commented on server/src/main/java/org/elasticsearch/cluster/metadata/MetadataIndexStateService.java:
```diff
@@ -147,114 +153,153 @@ public MetadataIndexStateService(
+                    state = addIndexClosedBlocks(task.request.indices(), blockedIndices, state);
+                    builder.success(task, task.listener.delegateFailure((delegate1, clusterState) -> {
+                        if (blockedIndices.isEmpty()) {
+                            delegate1.onResponse(new CloseIndexResponse(true, false, List.of()));
```
aa32395

elasticmachine commented:
Pinging @elastic/es-data-management (Team:Data Management)
